### PR TITLE
Add and update the docs for  "Defining a Schema" and "Fetching Data"

### PR DESCRIPTION
**Pull request by rkoron007**
This PR updates the articles for the "Defining a Schema" and "Fetching Data" docs sections (minus the [Data sources](https://www.apollographql.com/docs/apollo-server/data/data-sources) article, which I'll update once data sources are working)

[Accompanying updated docs-examples](https://github.com/apollographql/docs-examples/pull/47)

Related to: https://github.com/apollographql/apollo-server/issues/6758

### CODE REVIEW

**Comment by StephenBarlow** [comment]
Hm, currently it's challenging to pick out the parts of this PR that differ from the AS3 docs, versus what's just being re-added verbatim.

Other reviewers might have a thought here, but what if, before looking at this PR, we merged a _different_ PR that just added these articles to the AS4 docs _exactly_ as they're written in AS3? That would cause the diff for _this_ PR to be limited to the portions that actually changed for v4

**Comment by rkoron007** [comment]
You make a great point @StephenBarlow!

> =what if, before looking at this PR, we merged a _different_ PR that just added these articles to the AS4 docs _exactly_ as they're written in AS3? That would cause the diff for _this_ PR to be limited to the portions that actually changed for v4

The only thing that would worry me about the above approach is that folks _could_ navigate to the outdated AS3 docs on the V4 branch of the docsite if they enter the direct URL. But, I feel like that's fine 🤷‍♀️. @trevor-scheer and @glasser  thoughts? I can handle merging in those v3 docs if we feel this is the way to go!

**Comment by rkoron007** [comment]
I have a PR for adding the AS3 versions of these docs here: https://github.com/apollographql/apollo-server/pull/6778

@glasser, if it sounds good to you, I'll merge the above PR, making this PR easier to review.

**Comment by rkoron007** [comment]
Okay, @StephenBarlow and @glasser, this is ready for review now—I merged in the AS3 docs, so we now have the diffs between the two to go off! :)

File: docs/source/data/errors.mdx
Comment (glasser) [inline docs/source/data/errors.mdx:406]: > all errors are sent to Studio by default

This is a bit of an exaggeration. It's more like "operations sent to Studio as detailed traces contain every error". Not that literally every error gets sent, because not all operations are sent to Studio as traces at all. In fact, basically none are if you're not a paid user (each time Apollo Server starts up it might send one message with traces and the response to that trace will tell it to stop sending traces if you're not paid).

(Also I mention below that as of ultra recently the default is to mask all errors)

File: docs/source/data/errors.mdx
Comment (glasser) [inline docs/source/data/errors.mdx:417]: slight nuance (not new, just noticing it now): you can define it in the gateway if you're trying to rewrite stuff like parse and validation errors that happen in the gateway. it just won't see field errors.

File: docs/source/data/errors.mdx
Comment (glasser) [inline docs/source/data/errors.mdx:452]: probably best to replace "property" with "extensions" in this section, or to say "the message or extensions".  code isn't really a property, it's an extension.

maybe worth noting somewhere that the error as seen by rewriteError is always a GraphQLError and that you might want to use unwrapResolverError on it.

File: docs/source/data/errors.mdx
Comment (glasser) [inline docs/source/data/errors.mdx:498]: (this is now the default, ish)

File: docs/source/data/errors.mdx
Comment (glasser) [inline docs/source/data/errors.mdx:541]: (previous line) you need to set all headers using lowercase header names now (it will throw otherwise)

(this line) there's an extra `response?.result?.errors` thing here (the structure will be different for defer/stream responses, thus the new "result" bit; also, having a sub-object that's exactly "the thing that goes in the body of the response" seems useful).

File: docs/source/data/errors.mdx
Comment (glasser) [inline docs/source/data/errors.mdx:542]: i changed this to statusCode but changed it back to status today since that seemed like an unnecessary annoyance :)

File: docs/source/data/file-uploads.mdx
Comment (glasser) [inline docs/source/data/file-uploads.mdx:6]: Honestly I would just remove this file. I think that graphql-upload is essentially broken by design; people who want to use it are welcome to use graphql-upload's own docs.  Maybe we can add a small section in the CSRF prevention saying "by the way if you decide to use graphql-upload, you will find that it gets blocked by the CSRF Prevention feature; instead of disabling CSRF prevention (which is really helping you here!) pass the `{'Apollo-Require-Preflight': 'true'}` header to `createUploadLink`".

But otherwise I feel like one major version after we drop the direct integration is long enough to keep around the documentation of a bad feature that has mostly been responsible for security issues and where we've rewritten our own best practices to say not to use it at all.

File: docs/source/data/resolvers.mdx
Comment (glasser) [inline docs/source/data/resolvers.mdx:159]: that's the default listen argument, so you can just do `await startStandaloneServer(server)` (this applies in many places).

File: docs/source/data/errors.mdx
Comment (glasser) [inline docs/source/data/errors.mdx:214]: you're not using the first two imports

File: docs/source/data/resolvers.mdx
Comment (glasser) [inline docs/source/data/resolvers.mdx:378]: prettier? more like uglier

What mechanism are you using to run prettier (or a different formatter) btw? We have `*.md` and `*.mdx` in our prettierignore file — my VSCode is set up to not format-on-save. Though I like prettier and would be happier to add

```
!docs/source/**/*.md
!docs/source/**/*.mdx
```

to the bottom of that file (then run `npm run prettier-fix` once).

It also looks like if we add `"proseWrap": "never" to `.prettierrc.json5` then it [formats wide tables better](https://github.com/prettier/prettier/pull/5701).

File: docs/source/data/resolvers.mdx
Comment (glasser) [inline docs/source/data/resolvers.mdx:381]: Let's start calling this `contextValue`, which is the term used in `graphql-js` and is what we're now using wherever it's a field name on an object (vs `context` which is the function that produces it).  (In this case, it's just a positional argument so the name is something we make up in the docs.)

File: docs/source/data/resolvers.mdx
Comment (glasser) [inline docs/source/data/resolvers.mdx:421]: semicolon

File: docs/source/data/resolvers.mdx
Comment (glasser) [inline docs/source/data/resolvers.mdx:423]: maybe just move this into `resolvers` above so it's real code that builds

File: docs/source/data/resolvers.mdx
Comment (glasser) [inline docs/source/data/resolvers.mdx:438]: we should probably make explicit somewhere that `context` is expected to be an async function (in TS at least)

File: docs/source/data/resolvers.mdx
Comment (glasser) [inline docs/source/data/resolvers.mdx:445]: In fact, in AS4, the TS types *require* it to be an async function.

File: docs/source/data/resolvers.mdx
Comment (glasser) [inline docs/source/data/resolvers.mdx:467]: might also just say that a resolver can be an async function

**Review by glasser — COMMENTED** [review]
Here's the first few files.

File: docs/source/data/subscriptions.mdx
Comment (glasser) [inline docs/source/data/subscriptions.mdx:6]: While we probably should change how it is described, we probably should continue to open with a note that Apollo Server does not have built-in support for subscriptions.

File: docs/source/data/subscriptions.mdx
Comment (glasser) [inline docs/source/data/subscriptions.mdx:95]: this is now more like "if you `app.use` expressMiddleware at a different path"

File: docs/source/data/subscriptions.mdx
Comment (glasser) [inline docs/source/data/subscriptions.mdx:132]: Do you think it's actually useful that we walk through the code bit by bit and then show it all again here? Maybe just showing it once (fully complete) and not walking through (with extra comments added) would be better?

File: docs/source/data/subscriptions.mdx
Comment (glasser) [inline docs/source/data/subscriptions.mdx:248]: (scroll up a few lines)

> Apollo Server uses a **publish-subscribe** (**pub/sub**) model to track events that update active subscriptions

we should probably remove all things that claim that *Apollo Server* does anything relating to subscriptions (should have been done w/ v3)

File: docs/source/data/subscriptions.mdx
Comment (glasser) [inline docs/source/data/subscriptions.mdx:701]: Perhaps this "migrating to graphql-ws" part could be removed and we can link to the v3 docs instead.

File: docs/source/schema/creating-directives.mdx
Comment (glasser) [inline docs/source/schema/creating-directives.mdx:19]: This is yet another file documenting somebody else's software. (In this case, software that used to be part of AS, then got factored out into a separate library, then the library got passed off to other people, and finally we removed the default integration.) I personally wish we didn't have so much of that, or that it was somewhere separate from the core AS docs. Maybe we consider leaving this file out for v4 and if there's popular demand we resurrect it?

**Review by glasser — COMMENTED** [review]
woo, schema part is (unsurprisingly) less changey

**Comment by glasser** [comment]
also, at a high level, most of the stuff in the schema/resolvers pages should ideally be TS-first with codegen! but fine if that's a separate project that comes after "restore and correct docs".

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx:44]: Oops, that first comma should have been omitted. So this should read as:

```
If Apollo Server recognizes your error type you can write custom logic to handle that error, like so:
```

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx:214]: Great point! I'll remove these

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx:351]: Thank you for suggesting! I added a section addressing this :)

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx:406]: I love the phrasing you suggested, so is now:

> If you connect Apollo Server to Studio, operations sent to Studio as detailed traces contain every error by default.

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx:408]: Thanks for the heads up! Updated this section and the following code sample

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx:417]: Cool! Re wrote this as:

> **For federated graphs**, define your `rewriteError` function in each subgraph's [inline trace plugin](./migration/rewriteEerror-plugin-option) to rewrite field errors. If you want to transform your gateway's parsing or validation errors, you can define your `rewriteError` function in your gateway.

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx:452]: Love it, followed both of these suggestions!

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx:406]: Wait, @glasser isn't this no longer true? Since the usage reporting plugin now omits [error details from traces by default](https://www.apollographql.com/docs/apollo-server/v4/migration#usage-reporting-and-inline-trace-plugins-mask-errors-by-default)?

Comment (glasser) [reply docs/source/data/errors.mdx:406]: Yeah, that's what I said in the second parenthetical paragraph.

My point is twofold:

- The original wording was just misleading even in the context of AS3: it implies *every* error generated gets sent to Studio, which has never been true.
- Also we mask it now

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx:498]: Tweaked the introduction to reflect this

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx:541]: Sweet, thank you!

File: docs/source/data/file-uploads.mdx
Comment (rkoron007) [reply docs/source/data/file-uploads.mdx:6]: Makes sense to me! I'll delete this file and make sure we add a note to the CORS article.

File: docs/source/data/resolvers.mdx
Comment (rkoron007) [reply docs/source/data/resolvers.mdx:159]: Great point, I'll go through and do that across the board for consistency

File: docs/source/data/resolvers.mdx
Comment (rkoron007) [reply docs/source/data/resolvers.mdx:381]: We definitely could do this, but it'd be something we'd want to change **everywhere** (including Odyssey's courses) for consistency. 

I will say I do like the name `context` for this positional argument because it _does_ make you think about the `context` initialization function for your server. So that logic leap between "how does my server know this information" feels a little shorter. 

However, if you feel like `contextValue` is the better name and makes more sense, then we can do that!

Comment (rkoron007) [reply docs/source/data/resolvers.mdx:381]: @trevor-scheer and @StephenBarlow just in case either of you have thoughts on this too.

File: docs/source/data/resolvers.mdx
Comment (rkoron007) [reply docs/source/data/resolvers.mdx:378]: 😂  Oh geez I didn't realize it'd get this bad! My VSCode prettier settings are now set to not format Markdown files, thank you for catching this!

File: docs/source/data/resolvers.mdx
Comment (rkoron007) [reply docs/source/data/resolvers.mdx:445]: Added a mention to this effect above and changed this to:

> Because the `context` initialization function is asynchronous, you can use it to establish database connections and wait for other operations to complete:

File: docs/source/data/resolvers.mdx
Comment (rkoron007) [reply docs/source/data/resolvers.mdx:467]: I tweaked this to:

> Resolvers can be asynchronous and perform async actions, such as fetching from a database or back-end API. To support this, a resolver can return a promise that resolves to any other supported return type.

File: docs/source/data/subscriptions.mdx
Comment (rkoron007) [reply docs/source/data/subscriptions.mdx:132]: Hmmm, it does feel like a bit much, but on the other hand, I could see this being confusing to implement, and at least this way, we **know** folks can't miss a step. I could go either way on this, honestly.

File: docs/source/data/subscriptions.mdx
Comment (rkoron007) [reply docs/source/data/subscriptions.mdx:248]: Great point! Tweaked this to:

> You can use the **publish-subscribe** (**pub/sub**) model to track events that update active subscriptions.

File: docs/source/data/subscriptions.mdx
Comment (rkoron007) [reply docs/source/data/subscriptions.mdx:701]: Love this!

File: docs/source/schema/creating-directives.mdx
Comment (rkoron007) [reply docs/source/schema/creating-directives.mdx:19]: For this article, what if we left it out of the sidebar for now? So we'd still have the file on hand in case people want it, and folks searching for "creating directives" would still find something for their search, but the article is altogether less visible?

File: docs/source/data/resolvers.mdx
Comment (rkoron007) [reply docs/source/data/resolvers.mdx]: Done!

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx]: Sweet, thank you for the clarification!

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx:295]: I thought so too, but if you'd like me to show how to do this, I can add a quick demo!

Comment (glasser) [reply docs/source/data/errors.mdx:295]: Eh, I think it's OK not to.

File: docs/source/schema/creating-directives.mdx
Comment (glasser) [reply docs/source/schema/creating-directives.mdx:19]: That seems a bit confusing to me. Is this entirely about wanting to optimize SEO? Frankly I think people looking for documentation on this functionality of `@graphql-tools/schema` would be best served by finding the docs for `@graphql-tools/schema`.

File: docs/source/data/errors.mdx
Comment (glasser) [inline docs/source/data/errors.mdx:410]: Let's provide some justification for why this isn't just an annoyance — namely, you can look at fields like `path` on the `GraphQLError`.

File: docs/source/data/errors.mdx
Comment (glasser) [inline docs/source/data/errors.mdx]: It's `sendErrorsInTraces` (plural). (Many instances.)

You, uh, might want to fix the instances in migration.mdx while you're at it, oops.

File: docs/source/data/errors.mdx
Comment (glasser) [inline docs/source/data/errors.mdx:451]: This doesn't seem to match the code snippet after it?

**Review by glasser — CHANGES_REQUESTED** [review]

File: docs/source/data/errors.mdx
Comment (glasser) [reply docs/source/data/errors.mdx]: I'm still not sure this makes sense. I think esp in this introductory place it would make more sense to say something more like "you can look at the `code` to act differently depending on what kind of error it was; the enum lists the various errors produced by Apollo Server itself, and you can also make up your own codes" or something?

AS itself is basically never "recognizing" error codes, just producing them.

File: docs/source/data/errors.mdx
Comment (glasser) [inline docs/source/data/errors.mdx:54]: This section should probably be called "Built-in error codes" or something

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx]: Eagle eyes! I'll correct those across the board 👍

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx:451]: I clarified this by splitting it out:

> To send all errors to Studio you can pass `{ unmodified: true }` to `sendErrorsInTraces`, like so:

```ts {7}
new ApolloServer({
  // etc.
  plugins: [
    ApolloServerPluginUsageReporting({
      // If you pass unmodified: true to the usage reporting
      // plugin, Apollo Studio receives ALL error details
      sendErrorsInTraces { unmodified: true },
    }),
  ],
});
```

If you want to report specific errors or modify an error before reporting it, you can pass a function to the `sendErrorsInTraces.transform` option, like so:

```ts {4-6}
new ApolloServer({
  // etc.
  plugins: [
    ApolloServerPluginUsageReporting({
      sendErrorsInTraces {
        transform: (err) => {
          if (err.extensions.code === 'MY_CUSTOM_CODE') {
            // returning null will skip reporting this error 
            return null;
          }

          // All other errors are reported.
          return err;
        },
      },
    }),
  ],
});
```

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx:54]: I like that! Renaming it :)

File: docs/source/data/resolvers.mdx
Comment (StephenBarlow) [reply docs/source/data/resolvers.mdx]: I agree it's a deceptively heavy lift of a change, but fortunately still not _tremendously_ heavy since the name of the parameter doesn't really ever matter since you can technically name it whatever you want.

I think I do lean more in favor of keeping `context`, just because it's a map of many properties and values, and `contextValue` might suggest something more "scalar"

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx:410]: Great idea! I added some background here:

> Note, if a _resolver_ throws the error, a `GraphQLError` is wrapped around the initially thrown error. This `GraphQLError` neatly formats the error and contains useful fields, such as the `path` where the error occurred.
>
> In you want to remove the outer `GraphQLError` to access the originally thrown error you can use `unwrapResolverError` from `@apollo/server/errors`. The `unwrapResolverError` function can remove the `GraphQLError` wrapping from a resolver error or return the error unaltered if it isn't from a resolver.
>
> So, we can rewrite the above code snippet to work for errors thrown in and outside of resolvers, like so:

File: docs/source/data/errors.mdx
Comment (rkoron007) [reply docs/source/data/errors.mdx]: Gotcha, I see what you mean. 

How about something like:

> You can check an error's `code` to determine why an error occurred and also add logic to respond to different types of errors, like so:

File: docs/source/schema/creating-directives.mdx
Comment (rkoron007) [reply docs/source/schema/creating-directives.mdx:19]: It's more like I already put in the effort to update this article and the corresponding example, so the lift has _already_ been done on our side of things. So if that can help people searching in our documentation, that'd be nice? I can remove it if you feel like it doesn't add value though!

I could alternatively remove this full article, but add a plug for the `docs-example` in the main Directives article for those interested? 

I don't have strong feelings here, so I'm down to do whatever you think is best!

File: docs/source/data/resolvers.mdx
Comment (glasser) [reply docs/source/data/resolvers.mdx]: contextValue is the term we're going to be using in our API where there's actually a field, and it's also the term used by the graphql-js API.

If you think that contextValue is a bad name for the context value, then let's consider backing out the AS4 change where we change every other use of it from `context` to `contextValue`... it's helpful feedback! But if we're making the distinction "`context` is a function returning `contextValue`" we should make it everywhere!

File: docs/source/migration.mdx
Comment (glasser) [inline docs/source/migration.mdx:958]: You lost colons in these two

**Review by glasser — APPROVED** [review]

File: docs/source/migration.mdx
Comment (rkoron007) [reply docs/source/migration.mdx]: Thank you for catching that, fixed!

**Pull request merged**
